### PR TITLE
Bysyncify support: Asyncify/Emterpreter-like functionality for the wasm backend

### PR DESCRIPTION
**Pull request by kripken**
Basically gets us up to parity with fastcomp in terms of async support. This support should be faster and better. However, it doesn't have whitelist/blacklist and coroutine support yet (unsure how important those are). But the main features all work:

  * Sleep support (`emscripten_sleep, SDL_Delay`, etc.)
  * Emscripten sync APIs (`emscripten_wget*` and others).
  * Synchronous fsync syscall.
  * Browser main loop integration.

This enables on the wasm backend a bunch of previously emterpreter-only tests.

cc @gabrielcuvillier - if you test this on Doom let me know how it works!

### CODE REVIEW

File: tests/runner.py [hunk 1]
```diff
@@ -1501,6 +1501,7 @@ def reftest(self, expected, manually_trigger=False):
   def check_btest_skip(self, args):
     if self.is_wasm_backend() and 'WASM=0' in args:
       self.skipTest('wasm2js does not yet support threads, dynamic linking, and some other features; skip browser testing for now')
+    pass
```
Comment (sbc100) [inline tests/runner.py]: This doesn't look like it does anything.

File: tests/test_browser.py [hunk 2]
```diff
@@ -2652,22 +2650,22 @@ def test_codemods(self):
...
+
   def test_wget_data(self):
-    with open('test.txt', 'w') as f:
-      f.write('emscripten')
-    self.btest(path_from_root('tests', 'test_wget_data.c'), expected='1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O2', '-g2'])
-    self.btest(path_from_root('tests', 'test_wget_data.c'), expected='1', args=['-s', 'EMTERPRETIFY=1', '-s', 'EMTERPRETIFY_ASYNC=1', '-O2', '-g2', '-s', 'ASSERTIONS=1'])
+    create_test_file('test.txt', 'emscripten')
+    self.btest(path_from_root('tests', 'test_wget_data.c'), expected='1', args=['-O2', '-g2'] + self.get_async_args())
+    if not self.is_wasm_backend():
+      self.btest(path_from_root('tests', 'test_wget_data.c'), expected='1', args=['-O2', '-g2', '-s', 'ASSERTIONS=1'] + self.get_async_args())
```
Comment (sbc100) [inline tests/test_browser.py]: Comment/bug as to why this doesn't this work?

**Comment by sbc100** [comment]
Amazing work alon!   I didn't see the binaryen side yet but I guess that is where the heavy lifting is.

File: tests/test_browser.py (hunk 2 above)
Comment (kripken) [reply tests/test_browser.py]: I'll add a comment. basically this is extra testing for the emterpreter which has a very complex assertions mode. This tests it doesn't break anything. I think it's excessive for bysyncify, which doesn't have any special behavior with assertions.

**Comment by Keno** [comment]
How do I try this out? I installed latest emscripten/binaryen using emsdk, and just to test added a simple `emscripten_sleep` call in my code, as well as setting `-s BYSYNCIFY=1`. However, I'm seeing `TypeError: Module._bysyncify_stop_unwind is not a function` in the console. Is there some other setting I need to set?

**Comment by kripken** [comment]
@Keno That should work. But you may need to use the `bysyncify2` branch in this repo, and very latest `master` in binaryen, as I'm in the process of fixing and optimizing some final things.

If that doesn't fix things for you, does say `./tests/runner.py browser.test_async_iostream` work? If nothing obvious explains what you see, please share your full testcase, could be a bug.

I hope to finish this and have a summarizing blogpost out in a week or two.

**Comment by Keno** [comment]
I think the problem might be that I'm not using the LLVM wasm backend (which unfortunately doesn't work for me for reasons I don't quite understand).

**Comment by kripken** [comment]
Ah yes, it doesn't work without the wasm backend. I'll add an error for fastcomp.

Btw, please report anything that doesn't work in the wasm backend - it should be stable at this point, we intend to switch to it by default soon.

**Comment by Keno** [comment]
Yeah, will do. Right now the symptom is that mpfr fails to find gmp when I try to build both with the LLVM backend (worked fine with fastcomp). I'll dig one level deeper and then file an issue with whatever project is responsible.

**Comment by Keno** [comment]
Found my error. GMP was using a ranlib that doesn't understand wasm, causing the `.a` to not have a symbol table and `wasm-ld` to fail to find the symbols in it (but not otherwise complaining). Seems like that'll be a fairly common problem, so probably a good idea for `wasm-ld` to warn or error about.

**Comment by Keno** [comment]
Alright, after many hours and a bunch of fixes and workarounds up and down the tool chain, I've finally gotten everything built with the LLVM backend. Good news and bad news: The unwind works, and the timeout gets scheduled properly also. However, it errors during the rewind with a rather cryptic `RuntimeError: unreachable executed` at `bysyncify_start_rewind http://localhost:8888/hello.wasm:24290850`. I'll dig in further, but let me know if there's something obvious I need to look at.

**Comment by kripken** [comment]
@Keno thanks for testing!

I've been working on setting up fuzzing, and looks like there are some fuzz bugs. So maybe you're seeing one of those. However, another possibility is that the unreachable in those functions is basically an assertion for the data structure being valid. Are you using the normal emscripten code `handleSleep`? Or did you manually call `bysyncify_start_rewind` yourself?

**Comment by Keno** [comment]
I'm just using the regular emscripten_sleep for now, with no special handling anywhere else. I was gonna look at the data structure that's passed to rewind, but unfortunately doing so crashes Firefox and chrome won't even get that far due to stack size limits at the moment, so I'll have to probably resolve those issues, before I can continue debugging this.

**Comment by kripken** [comment]
I see, thanks @Keno. In that case maybe wait until I finish all the fuzz bugs, as then things should be a lot more stable. I can let you know when that is.

**Comment by kripken** [comment]
@Keno I fixed the fuzz issues I was seeing, and landed bysyncify2 meanwhile. So worth testing on latest binaryen master and emscripten incoming now, I think.

**Comment by Akaricchi** [comment]
How feasible would it be to implement coroutines? I'm currently experimenting with coroutines in my C game project, if I could keep emscripten compatibility that'd be awesome.

**Comment by kripken** [comment]
It's definitely possible to add, if someone is interested to do that. Bysyncify implements unwinding and rewinding, and for sleeping there is just one "context" that is unwound and rewound. For coroutines there would be multiple ones being juggled, and also the C stack may need to be saved, but that's mostly it I think. Here is where the emterpreter support was added, might be helpful (but the mechanisms are different), https://github.com/emscripten-core/emscripten/pull/6014

**Comment by Keno** [comment]
@kripken I've attempted to do some more debugging and I think my problem has to do with bysyncify unwinding through functions on the stack that are using `longjmp`. Is the expected to be supported?

**Comment by Keno** [comment]
In particular, if I'm interpreting what I'm seeing correctly, it looks like the binaryen pass doesn't resolve the `invoke_*` call targets, just treating them as imports not in the bysyncify import list and thus assuming that they cannot start an unwind.

**Comment by kripken** [comment]
Thanks @Keno, yeah, I forot that and I guess the emterpreter/asyncify tests didn't cover it. I'll have a fix soon.

**Comment by Keno** [comment]
Can confirm that things work with #8895. There's two remaining issues, though I can work around them on my side and run with this a bit more:
1. The unwind buffer is not checked against overflow (or if it's support to be that isn't working), so if it's too small (which the default is most of the time in my case), it goes around overwriting random portions of heap (and failing to rewind).
2. Activating bysyncify seems to blow up stack usage (?). In any case, Chrome starts throwing `RangeError: Maximum call stack size exceeded`, with very few (tens) stack frames on the stack when bysyncify is turned on. Not sure if this is from the extra frames introduced by instrumenting the exports of if Chrome measures something about the transformed wasm functions.

**Comment by kripken** [comment]
@Keno - do you have a testcase for the second issue? That seems very surprising, not sure what could cause that.

For the first issue, it should check for overflow (after unwinding completes), so sounds like something's wrong, I'll look into that.

**Pull request merged**